### PR TITLE
Use Weak pointers to automatically garbage collect old position maps

### PR DESCRIPTION
**Pull request by wz1000**
The size of the map does stop growing, but I can't seem to see much difference in the `-hT` profile.

I'll try to do some more detailed experiments soon.

### CODE REVIEW

**Comment by wz1000** [comment]
I don't know, we could theoretically have stale data referencing arbitrarily old versions.

**Comment by wz1000** [comment]
Before/after graph using `-hi` profiling:

Before

After

It is a small leak, but a leak nonetheless.

File: ghcide/src/Development/IDE/Types/Shake.hs [hunk 1]
```diff
@@ -31,10 +38,40 @@ import           Development.Shake                    (RuleResult,
...
 import           Language.LSP.Types
-
+import Control.Concurrent.MVar
+import Data.Void
+import System.Mem.Weak
+import Data.Coerce
+
+-- | A Token is a value with a finalizer. The finalizer is called when
+-- there are no more live references to the the token
+newtype Token = Token (MVar Void)
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Types/Shake.hs:48]: Why an `MVar` instead of an `IORef`? Which one is "smaller"?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 2]
```diff
@@ -175,7 +177,8 @@ data ShakeExtras = ShakeExtras
     ,publishedDiagnostics :: Var (HMap.HashMap NormalizedUri [Diagnostic])
     -- ^ This represents the set of diagnostics that we have published.
     -- Due to debouncing not every change might get published.
-    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping)))
+    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping, WeakToken)))
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Shake.hs:180]: How about 

```
type WeakValue a = (a, WeakToken)
```

```suggestion
    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (WeakValue (PositionDelta, PositionMapping))))
```

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 3]
```diff
@@ -1139,17 +1154,27 @@ filterVersionMap =
...
 updatePositionMapping :: IdeState -> VersionedTextDocumentIdentifier -> List TextDocumentContentChangeEvent -> IO ()
-updatePositionMapping IdeState{shakeExtras = ShakeExtras{positionMapping}} VersionedTextDocumentIdentifier{..} (List changes) = do
+updatePositionMapping IdeState{shakeExtras = ShakeExtras{positionMapping, liveTokens}} VersionedTextDocumentIdentifier{..} (List changes) = do
+    let uri = toNormalizedUri _uri
+    weak <- modifyVar liveTokens $ \oldToks -> do
+      let go x@(Just (_,ver,weak))
+            | ver > _version = Compose $ pure (weak, x)
+          go _ = Compose $ do
+            (tok, weak) <- newToken $ do
+              traceIO ("******************************************Finalized" ++ show (uri,_version))
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Shake.hs:1164]:
```suggestion
```

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 4]
```diff
@@ -175,7 +177,8 @@ data ShakeExtras = ShakeExtras
     ,publishedDiagnostics :: Var (HMap.HashMap NormalizedUri [Diagnostic])
     -- ^ This represents the set of diagnostics that we have published.
     -- Due to debouncing not every change might get published.
-    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping)))
+    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping, WeakToken)))
+    ,liveTokens :: Var (HMap.HashMap NormalizedUri (Token, TextDocumentVersion, WeakToken))
     -- ^ Map from a text document version to a PositionMapping that describes how to map
     -- positions in a version of that document to positions in the latest version
     -- First mapping is delta from previous version and second one is an
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Shake.hs:184]: comment is misplaced now

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 5]
```diff
@@ -175,7 +177,8 @@ data ShakeExtras = ShakeExtras
     ,publishedDiagnostics :: Var (HMap.HashMap NormalizedUri [Diagnostic])
     -- ^ This represents the set of diagnostics that we have published.
     -- Due to debouncing not every change might get published.
-    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping)))
+    ,positionMapping :: Var (HMap.HashMap NormalizedUri (Map TextDocumentVersion (PositionDelta, PositionMapping, WeakToken)))
+    ,liveTokens :: Var (HMap.HashMap NormalizedUri (Token, TextDocumentVersion, WeakToken))
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Shake.hs:181]: please add a comment

Comment (pepeiborra) [inline ghcide/src/Development/IDE/Core/Shake.hs:181]: I don't see the `liveTokens` map being pruned anywhere. How does this scheme work then?

**Review by pepeiborra — COMMENTED** [review]
Can we have some more comments across the code please? It's very hard to follow.

If I understand correctly, the tokens are used to track the liveness of `PositionMapping` values. Tokens are kept alive by stale `Value`s in the `values` mutable, and when they get collected their finaliser runs and prunes the `positionMapping` state.

But the `liveTokens` state seems to keep all the tokens alive, contradicting my understanding

**Comment by wz1000** [comment]
> But the `liveTokens` state seems to keep all the tokens alive, contradicting my understanding

`liveTokens` keeps one `Token` alive for the latest version of each document.

**Comment by pepeiborra** [comment]
How did you come up with this encoding of weak references?

**Comment by wz1000** [comment]
> How did you come up with this encoding of weak references? I've never managed to get them to work correctly in Haskell

I think worker-wrapper and other optimisations can destroy object identity very often, which means objects are finalized sooner than you would expect. But this shouldn't happen for objects with guaranteed identity, like `IORef` and `MVars`.

From the documentation: https://hackage.haskell.org/package/base-4.14.0.0/docs/System-Mem-Weak.html#t:Weak

> WARNING: weak pointers to ordinary non-primitive Haskell types are particularly fragile, because the compiler is free to optimise away or duplicate the underlying data structure. Therefore attempting to place a finalizer on an ordinary Haskell type may well result in the finalizer running earlier than you expected. This is not a problem for caches and memo tables where early finalization is benign.

> Finalizers can be used reliably for types that are created explicitly and have identity, such as IORef and MVar. However, to place a finalizer on one of these types, you should use the specific operation provided for that type, e.g. mkWeakIORef and addMVarFinalizer respectively (the non-uniformity is accidental). These operations attach the finalizer to the primitive object inside the box (e.g. MutVar# in the case of IORef), because attaching the finalizer to the box itself fails when the outer box is optimised away by the compiler.

I think `ByteArray#` will also work, I will try replacing the `MVar` with a zero-length `ByteArray#` tomorrow.

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 6]
```diff
@@ -931,20 +941,25 @@ defineEarlyCutoff op = addBuiltinRule noLint noIdentity $ \(Q (key, file)) (old
                 (bs, (diags, res)) <- actionCatch
                     (do v <- op key file; liftIO $ evaluate $ force v) $
                     \(e :: SomeException) -> pure (Nothing, ([ideErrorText file $ T.pack $ show e | not $ isBadDependency e],Nothing))
-                modTime <- liftIO $ (currentValue . fst =<<) <$> getValues state GetModificationTime file
```
Comment (wz1000) [inline ghcide/src/Development/IDE/Core/Shake.hs:934]: There is a race here, we shouldn't assign the latest version of the file as the version, rather we should use the `Token` corresponding to the version that is in the store.

However, this is a bit tricky since `getVirtualFile` returns just a `TextDocumentVersion` and not a `Token`. We also can't maintain a map of `Token`s to . To fix this, we would need to explicitly synchronize the lsp VFS with the document change notifications we define handlers for. At that point, we would have a token available, and we can return that along with the virtual file. This means that the type of `getVirtualFile` would change to `NormalizedUri -> IO (Maybe (VirtualFile, Token))`. 

I believe this bit was already somewhat racy(https://github.com/haskell/haskell-language-server/issues/1172), since `getVirtualFile` would always return the latest version of the document according to lsp, which might be ahead of the 'snapshot'  of the VFS we assume in the Shake sessioin.

We should maintain a separate snapshot of the VFS which is updated (via `LSP.getVirtualFile`) on `documentChange` events. This store can also keep track of the `Tokens` corresponding to the versions of the file in the VFS.

**Comment by wz1000** [comment]
Bonus before/after graphs:

2021-03-10-000345_grab

2021-03-10-000356_grab

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 2 above)
Comment (isovector) [reply ghcide/src/Development/IDE/Core/Shake.hs:180]:
```haskell
type WeakValue a = (WeakToken, a)
```

is better, so that it's a functor.

**Comment by Anton-Latukha** [comment]
This got blocked mostly by update to STM. I am probably not proficient enough to rebase it properly.

**Comment by Anton-Latukha** [comment]
I would assume to know what I am doing rebasing this.

**Comment by Anton-Latukha** [comment]
During current rebase there are indeed parts like:
```haskell
<<<<<<< HEAD
                modTime <- liftIO $ (currentValue . fst =<<) <$> atomicallyNamed "define - read 2" (getValues state GetModificationTime file)
=======
                live <- liftIO $ readVar liveTokens
                let !liveRes = HMap.lookup (filePathToUri' file) live
                    mtime = snd3 =<< liveRes
                    tok = fmap fst3 liveRes
                _ <- liftIO $ evaluate $ force (mtime,tok)
>>>>>>> 95c097c4 (Use Weak pointers to automatically garbage collect old position maps)
```
that need an understanding of the topic.

**Comment by michaelpj** [comment]
A lot has changed in memory usage, I guess that this no longer seems like a good avenue to investigate.
